### PR TITLE
Remove call into python API without GIL being held in c10d

### LINKED ISSUE #26475: c10d Python memory allocated without holding the GIL
## 🐛 Bug

Running the c10d test with a debug python 3.6 build leads to the following crash:
```
Fail to import hypothesis in common_utils, tests are not derandomized
.sssssssss....sssssssssssss...sssssss.sssFatal Python error: Python memory allocator called without holding the GIL

Current thread 0x00007fb16dff3080 (most recent call first):
  File "test_c10d.py", line 204 in _test_set_get
  File "test_c10d.py", line 210 in test_set_get
  File "/my_home/python3.6/debug/install/lib/python3.6/unittest/case.py", line 605 in run
  File "/my_home/python3.6/debug/install/lib/python3.6/unittest/case.py", line 653 in __call__
  File "/my_home/python3.6/debug/install/lib/python3.6/unittest/suite.py", line 122 in run
  File "/my_home/python3.6/debug/install/lib/python3.6/unittest/suite.py", line 84 in __call__
  File "/my_home/python3.6/debug/install/lib/python3.6/unittest/runner.py", line 176 in run
  File "/my_home/python3.6/debug/install/lib/python3.6/unittest/main.py", line 256 in runTests
  File "/my_home/python3.6/debug/install/lib/python3.6/unittest/main.py", line 95 in __init__
  File "/my_home/pytorch_3.6_debug_source/test/common_utils.py", line 117 in run_tests
  File "test_c10d.py", line 3080 in <module>
Aborted (core dumped)
```

The extra checks of the debug build seem to find a problem with the way the GIL is handled in c10d.

## To Reproduce

1. Install a python 3.6 debug build
2. Install pytorch from source using master
3. run c10d test with `python test_c10d.py`

### PR DESCRIPTION
**Pull request by albanD**
Fix #26475

### CODE REVIEW

**Comment by albanD** [comment]
The lint failure is from master.

File: torch/csrc/distributed/c10d/init.cpp [hunk 1]
```diff
@@ -650,11 +650,13 @@ Example::
           .def(
               "get",
               [](::c10d::Store& store, const std::string& key) -> py::bytes {
-                auto value = store.get(key);
+                auto value = [&]() {
+                  py::gil_scoped_release guard;
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/init.cpp:654]: Can you do what's being done on L627 with `py::call_guard<py::gil_scoped_release>(),` outside of the function implementation?

**Review by rohan-varma — APPROVED** [review]
LGTM just one minor comment

File: torch/csrc/distributed/c10d/init.cpp (hunk 1 above)
Comment (albanD) [reply torch/csrc/distributed/c10d/init.cpp:654]: It is fine to use as long as you don't call into python API inside the lambda.
The problem here is only the call to `py::bytes()` as this creates a new python object and thus must have the GIL held.
This is why the fix here is to restrict the guard to only apply to your c++ function and not to the py::bytes call.

**Pull request closed without merging**